Re-implement Identify Cluster as CodeDriven

soares-sergio opened the pull request:
#### Summary

This PR contains a code-driven C++ implementation of the Matter Identify cluster server. This implementation (`IdentifyCluster.h`) is designed for flexibility avoiding the the tight coupling present in older ZAP/Ember based implementations.

It uses a delegate pattern (chip::app::Clusters::IdentifyDelegate) to notify the application about cluster-related events, such as when identification starts, stops, or an effect is triggered.

#### Documentation

Added a README.md with documentation.

#### Testing

Added a comprehensive set of unit tests.

```
$ ninja -C out src/app/clusters/identify-server/tests:TestIdentifyCluster && ./out/tests/TestIdentifyCluster
ninja: Entering directory `out'
[3/3] ld tests/TestIdentifyCluster
[==========] Running all tests.
[ RUN      ] TestIdentifyCluster.TestCreate
[       OK ] TestIdentifyCluster.TestCreate
[ RUN      ] TestIdentifyCluster.AttributeListTest
[       OK ] TestIdentifyCluster.AttributeListTest
[ RUN      ] TestIdentifyCluster.AcceptedCommandsMandatoryOnlyTest
[       OK ] TestIdentifyCluster.AcceptedCommandsMandatoryOnlyTest
[ RUN      ] TestIdentifyCluster.AcceptedCommandsWithOptionalTriggerEffectTest
[       OK ] TestIdentifyCluster.AcceptedCommandsWithOptionalTriggerEffectTest
[ RUN      ] TestIdentifyCluster.ClusterRevisionTest
[       OK ] TestIdentifyCluster.ClusterRevisionTest
[ RUN      ] TestIdentifyCluster.FeatureMapTest
[       OK ] TestIdentifyCluster.FeatureMapTest
[ RUN      ] TestIdentifyCluster.ReadIdentifyAttributesTest
[       OK ] TestIdentifyCluster.ReadIdentifyAttributesTest
[ RUN      ] TestIdentifyCluster.WriteIdentifyTimeTest
[       OK ] TestIdentifyCluster.WriteIdentifyTimeTest
[ RUN      ] TestIdentifyCluster.WriteReadOnlyAttributesReturnUnsupportedWriteTest
[       OK ] TestIdentifyCluster.WriteReadOnlyAttributesReturnUnsupportedWriteTest
[ RUN      ] TestIdentifyCluster.IdentifyTypeCtorInitTest
[       OK ] TestIdentifyCluster.IdentifyTypeCtorInitTest
[ RUN      ] TestIdentifyCluster.IdentifyTimeCountdownTest
[       OK ] TestIdentifyCluster.IdentifyTimeCountdownTest
[ RUN      ] TestIdentifyCluster.OnIdentifyStartStopCallbackTest
[       OK ] TestIdentifyCluster.OnIdentifyStartStopCallbackTest
[ RUN      ] TestIdentifyCluster.OnStartNotCalledMultipleTimes
[       OK ] TestIdentifyCluster.OnStartNotCalledMultipleTimes
[ RUN      ] TestIdentifyCluster.OnStopNotCalledIfNotIdentifying
[       OK ] TestIdentifyCluster.OnStopNotCalledIfNotIdentifying
[ RUN      ] TestIdentifyCluster.InvokeIdentifyCommandTest
[       OK ] TestIdentifyCluster.InvokeIdentifyCommandTest
[ RUN      ] TestIdentifyCluster.InvokeTriggerEffectCommandTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0x0 variant 0x0
[       OK ] TestIdentifyCluster.InvokeTriggerEffectCommandTest
[ RUN      ] TestIdentifyCluster.InvokeTriggerEffectCommandAllEffectsTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0x1 variant 0x0
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0x2 variant 0x0
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0xB variant 0x0
[       OK ] TestIdentifyCluster.InvokeTriggerEffectCommandAllEffectsTest
[ RUN      ] TestIdentifyCluster.InvokeTriggerEffectCommandInvalidVariantTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0x0 variant 0x1
[       OK ] TestIdentifyCluster.InvokeTriggerEffectCommandInvalidVariantTest
[ RUN      ] TestIdentifyCluster.TriggerEffectWhileIdentifyingTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0x0 variant 0x0
[       OK ] TestIdentifyCluster.TriggerEffectWhileIdentifyingTest
[ RUN      ] TestIdentifyCluster.TriggerEffectFinishEffectTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0xFE variant 0x0
[       OK ] TestIdentifyCluster.TriggerEffectFinishEffectTest
[ RUN      ] TestIdentifyCluster.TriggerEffectStopEffectTest
[1759442143.167] [1818182:1818182] [ZCL] RX identify:trigger effect identifier 0xFF variant 0x0
[       OK ] TestIdentifyCluster.TriggerEffectStopEffectTest
[ RUN      ] TestIdentifyCluster.IdentifyTimeAttributeReportingTest
[       OK ] TestIdentifyCluster.IdentifyTimeAttributeReportingTest
[==========] Done running all tests.
[  PASSED  ] 22 test(s).
```

Backwards compatibility tests:

```
$ ninja -C out src/app/clusters/identify-server/tests:TestIdentifyClusterBackwardsCompatibility && ./out/tests/TestIdentifyClusterBackwardsCompatibility
ninja: Entering directory `out'
ninja: no work to do.
[==========] Running all tests.
[ RUN      ] TestIdentifyClusterBackwardsCompatibility.TestLegacyInstantiattion
[       OK ] TestIdentifyClusterBackwardsCompatibility.TestLegacyInstantiattion
[ RUN      ] TestIdentifyClusterBackwardsCompatibility.TestLegacyCallbacks
[1759442469.598] [1822133:1822133] [ZCL] RX identify:trigger effect identifier 0x0 variant 0x0
[       OK ] TestIdentifyClusterBackwardsCompatibility.TestLegacyCallbacks
[ RUN      ] TestIdentifyClusterBackwardsCompatibility.TestCurrentEffectIdentifierUpdate
[1759442469.598] [1822133:1822133] [ZCL] RX identify:trigger effect identifier 0x0 variant 0x0
[1759442469.598] [1822133:1822133] [ZCL] RX identify:trigger effect identifier 0xFE variant 0x0
[       OK ] TestIdentifyClusterBackwardsCompatibility.TestCurrentEffectIdentifierUpdate
[ RUN      ] TestIdentifyClusterBackwardsCompatibility.TestIdentifyTypeInitialization
[       OK ] TestIdentifyClusterBackwardsCompatibility.TestIdentifyTypeInitialization
[==========] Done running all tests.
[  PASSED  ] 4 test(s).
```


## Code Size Overhead (300 bytes or 600 bytes)

The average code size overhead is in the order of ~600 bytes if the application uses the legacy API.

I've updated the `all-clusters-app` for ESP32 to use the new CodeDriven API and instantiate the cluster directly, removing the CodegenIntegration (backwards compatibility layer) overhead. This makes the code size go down to ~300 bytes.
I also added a readme file describing how to migrate legacy apps to the new API and avoid the overhead.

Here's a size diff demonstrating the 300 bytes overhead:

```
$ ./scripts/tools/binary_elf_size_diff.py identify-chip-all-clusters-app.elf master-chip-all-clusters-app.elf 
Type       Size  Function                                                                             Size1       Size2
-------  ------  ------------------------------------------------------------------------------  ----------  ----------
REMOVED    -218  MatterIdentifyClusterServerAttributeChangedCallback                                      0         218
REMOVED    -152  chip::app::Clusters::Identify::DispatchServerCommand                                     0         152
REMOVED    -146  emberAfIdentifyClusterTriggerEffectCallback                                              0         146
REMOVED    -100  OnIdentifyTriggerEffect                                                                  0         100
REMOVED     -96  _GLOBAL__sub_I__Z23OnIdentifyTriggerEffectP8Identify                                     0          96
REMOVED     -64  onIdentifyClusterTick                                                                    0          64
REMOVED     -54  chip::app::Clusters::Identify::Attributes::IdentifyTime::Set                             0          54
REMOVED     -48  emberAfIdentifyClusterIdentifyCallback                                                   0          48
REMOVED     -46  Identify::Identify                                                                       0          46
REMOVED     -40  chip::app::Clusters::Identify::Attributes::IdentifyTime::Get                             0          40
REMOVED     -38  emberAfIdentifyClusterServerInitCallback                                                 0          38
CHANGED     -36  attributeData                                                                         4272        4308
REMOVED     -30  chip::app::Clusters::Identify::Attributes::IdentifyTime::Set                             0          30
REMOVED     -30  chip::app::Clusters::Identify::Attributes::IdentifyType::Set                             0          30
REMOVED     -28  gIdentify1                                                                               0          28
REMOVED     -28  __global_pointer$                                                                        0          28
REMOVED     -28  gIdentify0                                                                               0          28
REMOVED     -26  inst                                                                                     0          26
CHANGED     -18  chip::app::DispatchSingleClusterCommand                                                488         506
REMOVED     -16  gIdentify0::{lambda                                                                      0          16
REMOVED     -16  gIdentify0::{lambda                                                                      0          16
REMOVED     -16  gIdentify1::{lambda                                                                      0          16
REMOVED     -16  gIdentify0::{lambda                                                                      0          16
REMOVED     -16  gIdentify0::{lambda                                                                      0          16
REMOVED     -16  gIdentify1::{lambda                                                                      0          16
REMOVED     -16  gIdentify1::{lambda                                                                      0          16
REMOVED     -16  gIdentify1::{lambda                                                                      0          16
REMOVED     -12  identify_deactivate                                                                      0          12
REMOVED      -8  (anonymous namespace)::chipFuncArrayIdentifyServer                                       0           8
CHANGED      -4  emberAfGroupsClusterAddGroupIfIdentifyingCallback                                      102         106
REMOVED      -4  firstIdentify                                                                            0           4
CHANGED       2  AppTask::Init                                                                          422         420
ADDED         4  non-virtual thunk to chip::app::Clusters::IdentifyCluster::~IdentifyCluster              4           0
ADDED         4  (anonymous namespace)::sIdentifyDelegate                                                 4           0
ADDED         4  (anonymous namespace)::IdentifyDelegateImpl::IsTriggerEffectEnabled() const              4           0
ADDED         4  (anonymous namespace)::sTimerDelegate                                                    4           0
ADDED         4  (anonymous namespace)::firstLegacyIdentify                                               4           0
ADDED         6  non-virtual thunk to chip::app::Clusters::IdentifyCluster::TimerFired                    6           0
ADDED         8  chip::app::Clusters::IdentifyCluster::~IdentifyCluster                                   8           0
ADDED        10  (anonymous namespace)::IdentifyDelegateImpl::~IdentifyDelegateImpl                      10           0
ADDED        16  chip::app::Clusters::Identify::Attributes::kMandatoryMetadata                           16           0
ADDED        16  (anonymous namespace)::IdentifyDelegateImpl::OnIdentifyStop                             16           0
ADDED        16  chip::app::Clusters::(anonymous namespace)::kAcceptedCommandsWithTriggerEffect          16           0
CHANGED      16  MatterClusterServerShutdownCallback                                                    212         196
CHANGED      16  MatterClusterServerInitCallback                                                        212         196
ADDED        16  (anonymous namespace)::IdentifyDelegateImpl::OnIdentifyStart                            16           0
ADDED        18  FindIdentifyClusterOnEndpoint                                                           18           0
ADDED        32  vtable for (anonymous namespace)::IdentifyDelegateImpl                                  32           0
ADDED        32  chip::app::Clusters::IdentifyCluster::Attributes                                        32           0
ADDED        44  MatterIdentifyClusterShutdownCallback                                                   44           0
ADDED        44  MatterIdentifyClusterInitCallback                                                       44           0
ADDED        48  gIdentifyCluster0                                                                       48           0
ADDED        48  gIdentifyCluster1                                                                       48           0
ADDED        52  chip::app::Clusters::IdentifyCluster::TimerFired                                        52           0
ADDED        62  AppInit                                                                                 62           0
ADDED        66  chip::app::Clusters::IdentifyCluster::AcceptedCommands                                  66           0
ADDED        70  (anonymous namespace)::GetLegacyIdentifyInstance                                        70           0
ADDED        88  chip::app::Clusters::IdentifyCluster::IdentifyCluster                                   88           0
ADDED        92  vtable for chip::app::Clusters::IdentifyCluster                                         92           0
ADDED       100  (anonymous namespace)::IdentifyDelegateImpl::OnTriggerEffect                           100           0
ADDED       112  chip::app::Clusters::IdentifyCluster::WriteAttribute                                   112           0
ADDED       118  _GLOBAL__sub_I_gIdentifyCluster0                                                       118           0
ADDED       138  chip::app::Clusters::IdentifyCluster::ReadAttribute                                    138           0
ADDED       142  chip::app::Clusters::IdentifyCluster::SetIdentifyTime                                  142           0
ADDED       238  chip::app::Clusters::IdentifyCluster::InvokeCommand                                    238           0
TOTAL       304                                                                                  1345451564  1345451260
```

soares-sergio committed:
Add initial CodeDriven IdentifyCluster implementation with unit tests

soares-sergio committed:
Integrate new Identify codedriven cluster into codegen and add backwards compat layer

soares-sergio committed:
Merge branch 'master' into identify_cluster

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

soares-sergio committed:
Link codegen mock model for tests backwards compat tests

soares-sergio committed:
Fix build issues

andy31415 commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+    switch (request.path.mAttributeId)
+    {
+    case Attributes::IdentifyTime::Id: {
+        mPreviousIdentifyTime = mIdentifyTime;
```
You need to return a "no write" result when identify time is unchanged.

https://github.com/project-chip/connectedhomeip/blob/master/src/app/data-model-provider/ActionReturnStatus.h#L51

andy31415 commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+void IdentifyCluster::IdentifyTimeAttributeChanged(bool isWrittenByClientOrCmd)
+{
+    // Start Identify
+    if (mPreviousIdentifyTime == 0 && mIdentifyTime > 0)
```
This method overall feels odd:

- seems like a post-processing function that uses mPreviousIdentifyTime, which is required to be stored by the caller. I would prefer this method being a `SetIdetifyTime` and it can store previous time as a local variable. Saves some minor ram and makes the code clearer without imposing prerequisites on the caller.

- Logic for changed or not could be encapsulating by returning kWriteSuccessNoOp (https://github.com/project-chip/connectedhomeip/blob/master/src/app/data-model-provider/ActionReturnStatus.h#L51) if we want to indicate a "success but do not modify". Unsure if this makes the code more readable, however a single place for calling NotifyAttributeChanged seems desirable.

andy31415 commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+    {
+    case Identify::Commands::Identify::Id: {
+        Identify::Commands::Identify::DecodableType data;
+        if (data.Decode(input_arguments) != CHIP_NO_ERROR)
```
I think we should just `ReturnErrorOnFailure` on these. If decoding must return invalid command, it should be up to the command decoder to return a correct CHIP_GLOBAL_IM_ERROR

soares-sergio replied:
Fixed, used `ReturnErrorOnFailure` as suggested.

andy31415 commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+        return DataModel::ActionReturnStatus(Protocols::InteractionModel::Status::Success);
+    }
+    default:
+        return DefaultServerCluster::InvokeCommand(request, input_arguments, handler);
```
you can return the invalid command right away - defaultservercluster cannot know any more about this processing that the implementation. You are fully handling this and do not require the base class here or in other parts that you have the authoritative answer.

soares-sergio replied:
Done

soares-sergio committed:
Implemented code review suggestions

soares-sergio committed:
Merge branch 'master' into identify_cluster

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

soares-sergio commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+        mPreviousIdentifyTime = mIdentifyTime;
+        mIdentifyTime         = data.identifyTime;
+        IdentifyTimeAttributeChanged(/* isWrittenByClientOrCmd */ true);
+        return DataModel::ActionReturnStatus(Protocols::InteractionModel::Status::Success);
```
Removed casts, thanks!

soares-sergio commented on src/app/clusters/identify-server/IdentifyCluster.cpp:
```diff
@@ -0,0 +1,236 @@
+using namespace chip::app::Clusters::Identify;
+
+namespace {
+DefaultTimerDelegate sDefaultTimerDelegate;
```
Yup, Moved `sDefaultTimerDelegate` to CodegenIntegration and made timerDelegate a reference

soares-sergio commented on src/app/clusters/identify-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,184 @@
+        IntegrationDelegate integrationDelegate;
+        CodegenClusterIntegration::RegisterServer({ .endpointId                = endpointId,
+                                                    .clusterId                 = Clusters::Identify::Id,
+                                                    .fixedClusterInstanceCount = 1,
```
Got rid of the IntegrationDelegate entirely. thanks

andy31415 commented:
/gemini review

andy31415 commented on src/app/clusters/identify-server/tests/TestIdentifyCluster.cpp:
```diff
@@ -0,0 +1,573 @@
+
+namespace {
+
+bool onIdentifyStartCalled    = false;
```
using globals feels like a bit like it could cause ordering issues especially if some test fails. I don't think this is the case here as you are resetting them in startup, however wondering what would prevent us from building some RAII on this. And the reasoning seems to be the `on*` methods that just take in the cluster. If we would allow a `intptr_t` argument like generic C callbacks, we would allow a context or, maybe even better, if we would allow a Delegate pattern we could subclass it and make it work.

So could we:
  - Make the callbacks be actually a pure virtual class of IdentifyDelegate that has methods and RAAI this?
  - or (a bit uglier) make the callbacks take a intptr_t and still RAII this with propper casts?

andy31415 commented on src/app/clusters/identify-server/tests/TestIdentifyCluster.cpp:
```diff
@@ -0,0 +1,573 @@
+        onEffectIdentifierCalled = false;
+    }
+
+    chip::Test::TestServerClusterContext mContext;
```
We may want a context for every test that needs one: the test context contains several accumulators (e.g. for events) and starting fresh each time seems desirable.

soares-sergio replied:
Moved this to the individual test cases. Thanks!

andy31415 commented on src/app/clusters/identify-server/tests/TestIdentifyCluster.cpp:
```diff
@@ -0,0 +1,573 @@
+    // Read and verify FeatureMap
+    uint32_t featureMap;
+    EXPECT_EQ(ReadAttribute(cluster, { endpoint, Identify::Id, FeatureMap::Id }, featureMap), CHIP_NO_ERROR);
+    EXPECT_EQ(featureMap, 0u);
```
Please remove change detector tests. I believe this is one as well, let's delete it. This cluster has no features.

soares-sergio replied:
I think this test is useful. It test it's not returning error and feature map is 0, I believe that's the right check for a cluster that doesn't have features.

Anyways, I merged them all in a single test to make things easier to read.

soares-sergio commented on src/app/clusters/identify-server/tests/TestIdentifyCluster.cpp:
```diff
@@ -0,0 +1,573 @@
+    const auto identifyTimePath = ConcreteDataAttributePath(endpoint, Identify::Id, IdentifyTime::Id);
+
+    // Write a value to IdentifyTime and verify it was written.
+    EXPECT_EQ(WriteAttribute(cluster, identifyTimePath, 10u), CHIP_NO_ERROR);
```
There's already a more complete test that validates the change notifications `IdentifyTimeAttributeReportingTest`. I'll remove this one

soares-sergio committed:
Address code review suggestions; add IdentifyDelegate, remove redundant tests, .

restyled-commits committed:
Restyled by clang-format

soares-sergio commented on src/app/clusters/identify-server/tests/TestIdentifyCluster.cpp:
```diff
@@ -0,0 +1,573 @@
+    uint16_t clusterRevision;
+    EXPECT_EQ(ReadAttribute(cluster, { endpoint, Identify::Id, Globals::Attributes::ClusterRevision::Id }, clusterRevision),
+              CHIP_NO_ERROR);
+    EXPECT_EQ(clusterRevision, 5u);
```
Removed the check for v5 but kept the read as it checks it doesn't error out.

soares-sergio committed:
Fix gemini suggestions

soares-sergio committed:
Merge branch 'master' into identify_cluster

soares-sergio committed:
Fix compat issue for member mActive

soares-sergio committed:
Fix ambiguous type in test

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
Another missing typecast to fix esp32 build..

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
fix esp32 build issues

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
fix unit test

soares-sergio committed:
Zap regen all

soares-sergio committed:
Fix nxp build

soares-sergio commented:
Size diff on `all-clusters-app` for ESP32 on Oct 7:

```
$ ./scripts/tools/binary_elf_size_diff.py identify-chip-all-clusters-app.elf master-chip-all-clusters-app.elf 
Type       Size  Function                                                                             Size1       Size2
-------  ------  ------------------------------------------------------------------------------  ----------  ----------
REMOVED    -218  MatterIdentifyClusterServerAttributeChangedCallback                                      0         218
REMOVED    -152  chip::app::Clusters::Identify::DispatchServerCommand                                     0         152
REMOVED    -146  emberAfIdentifyClusterTriggerEffectCallback                                              0         146
REMOVED     -64  onIdentifyClusterTick                                                                    0          64
REMOVED     -54  chip::app::Clusters::Identify::Attributes::IdentifyTime::Set                             0          54
REMOVED     -48  emberAfIdentifyClusterIdentifyCallback                                                   0          48
REMOVED     -46  Identify::Identify                                                                       0          46
REMOVED     -38  emberAfIdentifyClusterServerInitCallback                                                 0          38
REMOVED     -30  chip::app::Clusters::Identify::Attributes::IdentifyTime::Set                             0          30
REMOVED     -30  chip::app::Clusters::Identify::Attributes::IdentifyType::Set                             0          30
REMOVED     -28  __global_pointer$                                                                        0          28
REMOVED     -26  inst                                                                                     0          26
CHANGED     -18  chip::app::DispatchSingleClusterCommand                                                488         506
REMOVED     -12  identify_deactivate                                                                      0          12
REMOVED      -8  (anonymous namespace)::chipFuncArrayIdentifyServer                                       0           8
REMOVED      -4  firstIdentify                                                                            0           4
ADDED         4  (anonymous namespace)::sDefaultTimerDelegate                                             4           0
ADDED         4  (anonymous namespace)::firstLegacyIdentify                                               4           0
ADDED         4  non-virtual thunk to chip::app::Clusters::IdentifyCluster::~IdentifyCluster              4           0
ADDED         4  (anonymous namespace)::IdentifyLegacyDelegate::IsTriggerEffectEnabled                    4           0
ADDED         4  (anonymous namespace)::gLegacyDelegate                                                   4           0
CHANGED       6  _GLOBAL__sub_I__Z23OnIdentifyTriggerEffectP8Identify                                   102          96
ADDED         6  (anonymous namespace)::IdentifyLegacyDelegate::~IdentifyLegacyDelegate                   6           0
ADDED         6  non-virtual thunk to chip::app::Clusters::IdentifyCluster::TimerFired                    6           0
ADDED         8  chip::app::Clusters::IdentifyCluster::~IdentifyCluster                                   8           0
CHANGED      16  MatterClusterServerInitCallback                                                        212         196
CHANGED      16  MatterClusterServerShutdownCallback                                                    212         196
ADDED        16  chip::app::Clusters::Identify::Attributes::kMandatoryMetadata                           16           0
ADDED        16  chip::app::Clusters::(anonymous namespace)::kAcceptedCommandsWithTriggerEffect          16           0
ADDED        32  chip::app::Clusters::IdentifyCluster::Attributes                                        32           0
ADDED        32  vtable for (anonymous namespace)::IdentifyLegacyDelegate                                32           0
ADDED        44  MatterIdentifyClusterShutdownCallback                                                   44           0
ADDED        44  MatterIdentifyClusterInitCallback                                                       44           0
CHANGED      44  gIdentify1                                                                              72          28
CHANGED      44  gIdentify0                                                                              72          28
ADDED        52  chip::app::Clusters::IdentifyCluster::TimerFired                                        52           0
ADDED        58  (anonymous namespace)::IdentifyLegacyDelegate::OnIdentifyStop                           58           0
ADDED        60  (anonymous namespace)::IdentifyLegacyDelegate::OnIdentifyStart                          60           0
ADDED        66  chip::app::Clusters::IdentifyCluster::AcceptedCommands                                  66           0
ADDED        70  (anonymous namespace)::GetLegacyIdentifyInstance                                        70           0
ADDED        78  (anonymous namespace)::IdentifyLegacyDelegate::OnTriggerEffect                          78           0
ADDED        88  chip::app::Clusters::IdentifyCluster::IdentifyCluster                                   88           0
ADDED        92  vtable for chip::app::Clusters::IdentifyCluster                                         92           0
ADDED       108  Identify::Identify                                                                     108           0
ADDED       136  chip::app::Clusters::IdentifyCluster::WriteAttribute                                   136           0
ADDED       138  chip::app::Clusters::IdentifyCluster::ReadAttribute                                    138           0
ADDED       142  chip::app::Clusters::IdentifyCluster::SetIdentifyTime                                  142           0
ADDED       238  chip::app::Clusters::IdentifyCluster::InvokeCommand                                    238           0
TOTAL       754                                                                                  1345452032  1345451278
```

soares-sergio commented on src/app/clusters/identify-server/CodegenIntegration.h:
```diff
@@ -0,0 +1,80 @@
+struct Identify
+{
+    /**
+     * Callbacks are not thread safe. To access the identify struct please
```
This comment about "callbacks not thread safe" was from the original code, so I left it here.

Also callbacks are not always called within chip timers as some events/callbacks are triggered by a command or attr write. I removed the comment to avoid confusion anyways.

soares-sergio committed:
More code review suggestions

soares-sergio committed:
Remove @file comments

soares-sergio commented on src/app/clusters/identify-server/IdentifyCluster.h:
```diff
@@ -0,0 +1,167 @@
+     * @brief Checks if the TriggerEffect command is enabled.
+     * @return True if the TriggerEffect command is enabled, false otherwise.
+     */
+    virtual bool IsTriggerEffectEnabled() = 0;
```
Yup, made const, thanks!

soares-sergio committed:
Zap regen after making Identify attributes AAI only

soares-sergio committed:
Merge branch 'master' into identify_cluster

soares-sergio committed:
Another regen after merge conflict fix

soares-sergio committed:
Merge branch 'master' into identify_cluster

soares-sergio committed:
decouble from groups server

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
Update all-clusters-app to instantiate Identify directly

soares-sergio committed:
Add readme

soares-sergio committed:
Merge branch 'master' into identify_cluster

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by prettier-markdown

soares-sergio committed:
Update readme with review suggestions

soares-sergio committed:
Attempt to fix CI - TestDescriptorCluster

soares-sergio committed:
Remove Identify from ep0 everywhere

soares-sergio committed:
merge master and regen

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by prettier-json

soares-sergio committed:
Fix CI issue for all-clusters missing Identify instances for EP2,3,4

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
make config constexpr

soares-sergio committed:
Propagate error on NotifyAttributeChangedIfSuccess

soares-sergio committed:
fix lock app identify on linux

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

andy31415 reviewed (approved):
On the assumption we fix build problems/deps this PR looks good: very solid unit tested, fully backwards API compatible with a README on upgrading and actually exposes oddities: some apps never have real identify, hence their flash decreases.

Flash-size wise I would say this ok:
  - larger increases include a compat layer
  - a test "remove compat layer" seems to be able to bring this down to O(200) bytes
  - flash decreases are odd showing that some platforms never really identify ... that is a bug and I believe it is up to the platform example to fix

soares-sergio committed:
Delete identify-server.cpp again...

soares-sergio committed:
Instantiate Identify on energy-management app (fix CI failure)

soares-sergio committed:
Merge branch 'master' into identify_cluster

soares-sergio committed:
Fix lit icd

soares-sergio committed:
Fix microwave app (instantiate Identify)

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

soares-sergio committed:
Fix rvc app

soares-sergio committed:
Fix light app

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

soares-sergio committed:
Merge remote-tracking branch 'upstream/master' into identify_cluster

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

soares-sergio committed:
Another fix for lit-icd includes

soares-sergio committed:
Fix darwin clang tidy nonsense

restyled-commits committed:
Restyled by clang-format

soares-sergio committed:
Fix clang tidy attempt 2...

soares-sergio committed:
Fix nitpicks on test code

restyled-commits committed:
Restyled by clang-format

soares-sergio commented on src/app/clusters/identify-server/tests/ClusterActions.h:
```diff
@@ -0,0 +1,111 @@
+    ReturnErrorOnFailure(readOperation.GetEncodedIBs().Decode(attributeData));
+    VerifyOrReturnError(attributeData.size() == 1u, CHIP_ERROR_INCORRECT_STATE);
+
+    return chip::app::DataModel::Decode(attributeData[0].dataReader, value);
```
Yup, good call. I renamed the function to `ReadNumericAttribute` and added a static assert to ensure it doesn't let anyone reuse this code and try to pass strings or lists.

jmartinez-silabs reviewed (approved):
shouldn't zap file be updated too to reflect the attributes now being external?

You can do `scripts/tools/zap_convert_all.py` to update them all. This pr is large already so It can be a follow-up
